pkg/prometheus_ingester/prometheus_ingester.go

```diff
@@ -75,9 +76,65 @@ type queryOptions struct {
 	ResultAsQuantity *bool
 }
 
+type httpHeaderValueFromEnv struct {
+	Name        string `yaml:"name"`
+	ValuePrefix string `yaml:"valuePrefix"`
+}
+
+type httpHeader struct {
+	Name         string
+	ValueFromEnv *httpHeaderValueFromEnv `yaml:"valueFromEnv"`
+	Value        *string                 `yaml:"value"`
+}
+
+func (h *httpHeader) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	// we have to declare new type in order to avoid infinit recursion when using unmarshal function
+	type httpHeaderCustom httpHeader
+
+	out := httpHeaderCustom{ValueFromEnv: &httpHeaderValueFromEnv{}}
```

Comment:
The "cleverness" to set one of the pointers to non-nil harms readability I think. Moreover, it makes the error message on line 107 confusing when the user does set `ValueFromEnv`, but doesn't set the `Name` there.

It is much easier to reason about the value of `out` if unmarshal is called on a zero-initialised struct. It is also possible to produce correct error messages.

I suggest `out` be zero-initialised, then deal with nil pointers and/or incorrectly set fields normally.